# Analyzing borrowers’ risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Your report will be considered when building the **credit score** of a potential customer. The **credit score** is used to evaluate the ability of a potential borrower to repay their loan.



## Open the data file and have a look at the general information. 


In [2]:
import pandas as pd
# Loading all the libraries

credit_scoring = pd.read_csv('/datasets/credit_scoring_eng.csv')
# Load the data


## Task 1. Data exploration

**Description of the data**
- `children` - the number of children in the family
- `days_employed` - work experience in days
- `dob_years` - client's age in years
- `education` - client's education
- `education_id` - education identifier
- `family_status` - marital status
- `family_status_id` - marital status identifier
- `gender` - gender of the client
- `income_type` - type of employment
- `debt` - was there any debt on loan repayment
- `total_income` - monthly income
- `purpose` - the purpose of obtaining a loan



In [3]:
credit_scoring.describe()
# Let's see how many rows and columns our dataset has



children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

In [4]:
credit_scoring.head(40)
# let's print the first N rows



children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42    bachelor's degree             0   
1          1   -4024.803754         36  secondary education             1   
2          0   -5623.422610         33  Secondary Education             1   
3          3   -4124.747207         32  secondary education             1   
4          0  340266.072047         53  secondary education             1   
5          0    -926.185831         27    bachelor's degree             0   
6          0   -2879.202052         43    bachelor's degree             0   
7          0    -152.779569         50  SECONDARY EDUCATION             1   
8          2   -6929.865299         35    BACHELOR'S DEGREE             0   
9          0   -2188.756445         41  secondary education             1   
10         2   -4171.483647         36    bachelor's degree             0   
11         0    -792.701887         40  secondary education             1   
12         0            NaN         65  secondary education             1   
13         0   -1846.641941         54         some college             2   
14         0   -1844.956182         56    bachelor's degree             0   
15         1    -972.364419         26  secondary education             1   
16         0   -1719.934226         35  secondary education             1   
17         0   -2369.999720         33    bachelor's degree             0   
18         0  400281.136913         53  secondary education             1   
19         0  -10038.818549         48  SECONDARY EDUCATION             1   
20         1   -1311.604166         36  secondary education             1   
21         1    -253.685166         33  secondary education             1   
22         1   -1766.644138         24  secondary education             1   
23         0    -272.981385         21    bachelor's degree             0   
24         1  338551.952911         57  secondary education             1   
25         0  363548.489348         67  secondary education             1   
26         0            NaN         41  secondary education             1   
27         0    -529.191635         28    bachelor's degree             0   
28         1    -717.274324         26    bachelor's degree             0   
29         0            NaN         63  secondary education             1   
30         1  335581.668515         62  secondary education             1   
31         0   -1682.083438         47    primary education             3   
32         0   -4649.910832         34  SECONDARY EDUCATION             1   
33         0   -1548.637544         48  secondary education             1   
34         0   -4488.067031         35  secondary education             1   
35         0  394021.072184         68  secondary education             1   
36         0    -176.216688         33  secondary education             1   
37         0   -6448.810860         43    BACHELOR'S DEGREE             0   
38         0    -597.881827         25    bachelor's degree             0   
39         0    -650.587796         31    bachelor's degree             0   

        family_status  family_status_id gender    income_type  debt  \
0             married                 0      F       employee     0   
1             married                 0      F       employee     0   
2             married                 0      M       employee     0   
3             married                 0      M       employee     0   
4   civil partnership                 1      F        retiree     0   
5   civil partnership                 1      M       business     0   
6             married                 0      F       business     0   
7             married                 0      M       employee     0   
8   civil partnership                 1      F       employee     0   
9             married                 0      M       employee     0   
10            married                 0      M       business    

There are missing values in the 'days_employed' and 'total_income' columns. There are also many negative values in the 'days_employed' column. Lastly, the 'days_employed' column should have int values instead of float values. 

In [5]:
credit_scoring.info()
# Get info on data


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


There are only missing values in the 'days_employed' and the 'total_income' columns. The number of missing values in both of these columns are equal and that would indicate that if one of those values are missing, the other one will be missing as well. 

In [6]:
# Let's look at the filtered table with missing values in the the first column with missing data

missing_data = credit_scoring.isna().sum()

missing_data

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Missing values seem to be symmetric. There are the exact same number of missing values for 'days_employed' and 'total_income'. 

In [7]:
# Let's apply multiple conditions for filtering data and look at the number of rows in the filtered table.

missing_data_rows = credit_scoring[(credit_scoring['days_employed'].isna()) & (credit_scoring['total_income'].isna())].count()

missing_data_rows

children            2174
days_employed          0
dob_years           2174
education           2174
education_id        2174
family_status       2174
family_status_id    2174
gender              2174
income_type         2174
debt                2174
total_income           0
purpose             2174
dtype: int64

**Intermediate conclusion**

Yes, the number of rows in the filtered table do match the number of missing values. This indicates that the only missing values in this dataset are the 2174 missing values in 'days_employed' and in 'total_income'.

In [8]:
# Checking distribution

credit_scoring.isna().sum()/len(credit_scoring) * 100


children             0.000000
days_employed       10.099884
dob_years            0.000000
education            0.000000
education_id         0.000000
family_status        0.000000
family_status_id     0.000000
gender               0.000000
income_type          0.000000
debt                 0.000000
total_income        10.099884
purpose              0.000000
dtype: float64

**Possible reasons for missing values in data**

If there is a missing datapoint in the 'days_employed' column, there will be a missing point in the total_income column as well. Besides that, there is no apparent pattern to the missing data.


**Conclusions**


No patterns were apparent in the missing data. There were no column values that were consistently the same throughout all of the missing data points.


The missing values are all in the 'days_employed' or 'total_income' categories. These values will be replaced with either the mean or median (depending on data distribution) of their respective columns after being grouped into relevant categories. 


Several factors will be addressed during data transformation. Education category will need to be fixed to have consistent language. There are nonsensical values in the 'children' category that need to be addressed. There are many negative values in the 'days_employed' category. There are several datapoints indicating an age of 0 that need to be addressed. There is an entry under gender that does not make sense. Lastly, there are many duplicates that need to be addressed. 

## Data transformation


In [9]:
# Let's see all values in education column to check if and what spellings will need to be fixed
credit_scoring['education'].value_counts()

secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             668
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
GRADUATE DEGREE            1
Graduate Degree            1
Name: education, dtype: int64

In [10]:
# Fix the registers if required
credit_scoring['education'] = credit_scoring['education'].str.lower()

In [11]:
# Checking all the values in the column to make sure we fixed them

credit_scoring['education'].value_counts()

secondary education    15233
bachelor's degree       5260
some college             744
primary education        282
graduate degree            6
Name: education, dtype: int64

In [12]:
# Let's see the distribution of values in the `children` column
credit_scoring['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

There are many instances of 20 children. While this is not completely unheard of, the fact that there are 76 cases of having 20 children and 0 cases between 5 and 20 is suspicious. This is most likely due to input error. These users most likely meant to input 2 instead of 20. There are also many cases of -1 children which is not possible. This is also most likely due to input error. These users most likely meant to enter 1.

In [13]:
# [fix the data based on your decision]
credit_scoring['children'] = credit_scoring['children'].replace([20, -1], [2, 1])



In [14]:
# Checking the `children` column again to make sure it's all fixed
credit_scoring['children'].value_counts()


0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

In [15]:
# Find problematic data in `days_employed`, if they exist, and calculate the percentage
credit_scoring['days_employed'].head()



0     -8437.673028
1     -4024.803754
2     -5623.422610
3     -4124.747207
4    340266.072047
Name: days_employed, dtype: float64

In [16]:
days_employed_negative = credit_scoring[credit_scoring['days_employed'] < 0]
days_employed_negative.count() / credit_scoring.count()

children            0.738955
days_employed       0.821973
dob_years           0.738955
education           0.738955
education_id        0.738955
family_status       0.738955
family_status_id    0.738955
gender              0.738955
income_type         0.738955
debt                0.738955
total_income        0.821973
purpose             0.738955
dtype: float64

It is not possible to be employed for a negative amount of days so all of the negative values in this column constitute problematic data. 73.9% of these values are problematic and they are most likely caused by entry error. These values will all be converted to positive numbers.

In [17]:
# Address the problematic values, if they exist
credit_scoring['days_employed'] = credit_scoring['days_employed'].abs()



In [18]:
# Check the result - make sure it's fixed
credit_scoring['days_employed'].head(40)



0       8437.673028
1       4024.803754
2       5623.422610
3       4124.747207
4     340266.072047
5        926.185831
6       2879.202052
7        152.779569
8       6929.865299
9       2188.756445
10      4171.483647
11       792.701887
12              NaN
13      1846.641941
14      1844.956182
15       972.364419
16      1719.934226
17      2369.999720
18    400281.136913
19     10038.818549
20      1311.604166
21       253.685166
22      1766.644138
23       272.981385
24    338551.952911
25    363548.489348
26              NaN
27       529.191635
28       717.274324
29              NaN
30    335581.668515
31      1682.083438
32      4649.910832
33      1548.637544
34      4488.067031
35    394021.072184
36       176.216688
37      6448.810860
38       597.881827
39       650.587796
Name: days_employed, dtype: float64

In [19]:
# Check the `dob_years` for suspicious values and count the percentage
credit_scoring['dob_years'].describe()

count    21525.000000
mean        43.293380
std         12.574584
min          0.000000
25%         33.000000
50%         42.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

In [20]:
credit_scoring['dob_years'].value_counts(normalize=True) *100

35    2.866434
40    2.829268
41    2.819977
34    2.801394
38    2.778165
42    2.773519
33    2.699187
39    2.662021
31    2.601626
36    2.578397
44    2.541231
29    2.531940
30    2.508711
48    2.499419
37    2.494774
50    2.387921
43    2.383275
32    2.369338
49    2.360046
28    2.336818
45    2.308943
27    2.290360
56    2.262485
52    2.248548
47    2.229965
54    2.225319
46    2.206736
58    2.141696
57    2.137050
53    2.132404
51    2.081301
59    2.062718
55    2.058072
26    1.895470
60    1.751452
25    1.658537
61    1.649245
62    1.635308
63    1.249710
64    1.231127
24    1.226481
23    1.180023
65    0.901278
66    0.850174
22    0.850174
67    0.775842
21    0.515679
0     0.469222
68    0.459930
69    0.394890
70    0.301974
71    0.269454
20    0.236934
72    0.153310
19    0.065041
73    0.037166
74    0.027875
75    0.004646
Name: dob_years, dtype: float64

There are many age values that were reported as 0 (0.46% of total values). These can most likely be attributed to entry error or lack of answer. These values should be replaced with the mean age.

In [21]:
# Address the issues in the `dob_years` column, if they exist
mean_age = credit_scoring['dob_years'].mean()

mean_age_int = mean_age.astype(int)

credit_scoring['dob_years'] = credit_scoring['dob_years'].replace(0, mean_age_int)

In [22]:
# Check the result - make sure it's fixed
(credit_scoring['dob_years'] == 0).sum()

0

In [23]:
# Let's see the values for the column

credit_scoring['family_status'].value_counts()

married              12380
civil partnership     4177
unmarried             2813
divorced              1195
widow / widower        960
Name: family_status, dtype: int64

There is one entry that does not have an answer for gender. Since this is less than 0.01% of the dataset, it is possible to just delete this row entirely without affecting the overall dataset.

In [24]:
# Let's see the values in the column

credit_scoring['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

In [25]:
# Address the problematic values, if they exist
gender_problem = credit_scoring[credit_scoring['gender'] == 'XNA'].index
credit_scoring.drop(gender_problem, inplace = True)

In [26]:
# Check the result - make sure it's fixed
credit_scoring['gender'].value_counts()


F    14236
M     7288
Name: gender, dtype: int64

In [27]:
# Let's see the values in the column

credit_scoring['income_type'].value_counts()

employee                       11119
business                        5084
retiree                         3856
civil servant                   1459
unemployed                         2
entrepreneur                       2
paternity / maternity leave        1
student                            1
Name: income_type, dtype: int64

Income will need to be categorized in order to ensure proper analysis.

In [28]:
def income_category(row):
    income = row['total_income']
    
    if 0 <= income <= 5000:
        return '0-5k'
    if 5000 < income <= 10000:
        return '5-10k'
    if 10000 < income <= 15000:
        return '10-15k'
    if 15000 < income <= 20000:
        return '15-20k'
    if 20000 < income <= 25000:
        return '20-25k'
    if 25000 < income <= 30000:
        return '25-30k'
    if 30000 < income <= 35000:
        return '30-35k'
    if 35000 < income <= 40000:
        return '35-40k'    
    if 40000 < income <= 45000:
        return '40-45k'
    if 45000 < income <= 50000:
        return '45-50k'
    if 50000 < income <= 55000:
        return '50-55k'
    if 55000 < income <= 60000:
        return '55-60k'
    return 'Over 60k'

credit_scoring['income_categories'] = credit_scoring.apply(income_category, axis = 1)

In [29]:
# Checking duplicates
credit_scoring.duplicated().value_counts()


False    21453
True        71
dtype: int64

In [30]:
# Address the duplicates, if they exist

credit_scoring = credit_scoring.drop_duplicates().reset_index(drop = True)

In [31]:
# Last check whether we have any duplicates
credit_scoring.duplicated().value_counts()


False    21453
dtype: int64

In [32]:
# Check the size of the dataset that you now have after your first manipulations with it

credit_scoring.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 13 columns):
children             21453 non-null int64
days_employed        19350 non-null float64
dob_years            21453 non-null int64
education            21453 non-null object
education_id         21453 non-null int64
family_status        21453 non-null object
family_status_id     21453 non-null int64
gender               21453 non-null object
income_type          21453 non-null object
debt                 21453 non-null int64
total_income         19350 non-null float64
purpose              21453 non-null object
income_categories    21453 non-null object
dtypes: float64(2), int64(5), object(6)
memory usage: 2.1+ MB


The number of rows in this dataset has decreased by 72 due to the duplicate rows that were deleted and the single row that was deleted for not having a gender value. These rows constituted 0.3% of the whole dataset.

# Working with missing values

In [33]:
# Find the dictionaries
education = {
    0 == 'bachelor\'s degree',
    1 == 'secondary education',
    2 == 'some college',
    3 =='primary education',
    4 == 'graduate degree'
}

family_status = {
    0 == 'married',
    1 == 'civil partnership',
    2 == 'widow / widower',
    3 == 'divorced',
    4 =='unmarried'
}

### Restoring missing values in `total_income`

The 'days_employed' and 'total_income' columns have missing data. In order to account for outlier data and achieve maximum accuracy, these values will be replaced with the median values of their respective columns according to the income type. 

In [34]:
# Let's write a function that calculates the age category
def age_category(row):
    age = row['dob_years']
    
    if 18 <= age <= 19:
        return 'Teen'
    if 20 < age <= 29:
        return '20\'s'
    if 30 < age <= 39:
        return '30\'s'
    if 40 < age <= 49:
        return '40\'s'
    if 50 < age <= 59:
        return '50\'s'
    if 60 < age <= 69:
        return '60\'s'
    return '70\'s'


In [35]:
# Test if the function works
row_value = [33]
row_column = ['dob_years']
row = pd.Series(data=row_value, index=row_column)
age_category(row)

"30's"

In [36]:
row_value = [70]
row_column = ['dob_years']
row = pd.Series(data=row_value, index=row_column)
age_category(row)

"70's"

In [37]:
# Creating new column based on function
credit_scoring['age_categories'] = credit_scoring.apply(age_category, axis = 1)


In [38]:
# Checking how values in the new column
credit_scoring['age_categories'].value_counts()


30's    5125
40's    4848
50's    4144
20's    3114
70's    2251
60's    1957
Teen      14
Name: age_categories, dtype: int64

In [39]:
credit_scoring.head(30)

children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42    bachelor's degree             0   
1          1    4024.803754         36  secondary education             1   
2          0    5623.422610         33  secondary education             1   
3          3    4124.747207         32  secondary education             1   
4          0  340266.072047         53  secondary education             1   
5          0     926.185831         27    bachelor's degree             0   
6          0    2879.202052         43    bachelor's degree             0   
7          0     152.779569         50  secondary education             1   
8          2    6929.865299         35    bachelor's degree             0   
9          0    2188.756445         41  secondary education             1   
10         2    4171.483647         36    bachelor's degree             0   
11         0     792.701887         40  secondary education             1   
12         0            NaN         65  secondary education             1   
13         0    1846.641941         54         some college             2   
14         0    1844.956182         56    bachelor's degree             0   
15         1     972.364419         26  secondary education             1   
16         0    1719.934226         35  secondary education             1   
17         0    2369.999720         33    bachelor's degree             0   
18         0  400281.136913         53  secondary education             1   
19         0   10038.818549         48  secondary education             1   
20         1    1311.604166         36  secondary education             1   
21         1     253.685166         33  secondary education             1   
22         1    1766.644138         24  secondary education             1   
23         0     272.981385         21    bachelor's degree             0   
24         1  338551.952911         57  secondary education             1   
25         0  363548.489348         67  secondary education             1   
26         0            NaN         41  secondary education             1   
27         0     529.191635         28    bachelor's degree             0   
28         1     717.274324         26    bachelor's degree             0   
29         0            NaN         63  secondary education             1   

        family_status  family_status_id gender    income_type  debt  \
0             married                 0      F       employee     0   
1             married                 0      F       employee     0   
2             married                 0      M       employee     0   
3             married                 0      M       employee     0   
4   civil partnership                 1      F        retiree     0   
5   civil partnership                 1      M       business     0   
6             married                 0      F       business     0   
7             married                 0      M       employee     0   
8   civil partnership                 1      F       employee     0   
9             married                 0      M       employee     0   
10            married                 0      M       business     0   
11            married                 0      F       employee     0   
12  civil partnership                 1      M        retiree     0   
13            married                 0      F       employee     0   
14  civil partnership                 1      F       business     1   
15            married                 0      F       employee     0   
16            married                 0      F       employee     0   
17  civil partnership                 1      M       employee     0   
18    widow / widower                 2      F        retiree     0   
19           divorced                 3      F       employee     0   
20            married                 0      M       employee     0   
21  civil partnership                 1      F       e

In [40]:
# Create a table without missing values and print a few of its rows to make sure it looks fine
credit_scoring_missfree = credit_scoring[credit_scoring['total_income'].isna()].index
credit_missing_none = credit_scoring.copy()
credit_missing_none.drop(credit_scoring_missfree, inplace=True)
credit_missing_none = credit_missing_none.reset_index()
credit_missing_none.head(30)


index  children  days_employed  dob_years            education  \
0       0         1    8437.673028         42    bachelor's degree   
1       1         1    4024.803754         36  secondary education   
2       2         0    5623.422610         33  secondary education   
3       3         3    4124.747207         32  secondary education   
4       4         0  340266.072047         53  secondary education   
5       5         0     926.185831         27    bachelor's degree   
6       6         0    2879.202052         43    bachelor's degree   
7       7         0     152.779569         50  secondary education   
8       8         2    6929.865299         35    bachelor's degree   
9       9         0    2188.756445         41  secondary education   
10     10         2    4171.483647         36    bachelor's degree   
11     11         0     792.701887         40  secondary education   
12     13         0    1846.641941         54         some college   
13     14         0    1844.956182         56    bachelor's degree   
14     15         1     972.364419         26  secondary education   
15     16         0    1719.934226         35  secondary education   
16     17         0    2369.999720         33    bachelor's degree   
17     18         0  400281.136913         53  secondary education   
18     19         0   10038.818549         48  secondary education   
19     20         1    1311.604166         36  secondary education   
20     21         1     253.685166         33  secondary education   
21     22         1    1766.644138         24  secondary education   
22     23         0     272.981385         21    bachelor's degree   
23     24         1  338551.952911         57  secondary education   
24     25         0  363548.489348         67  secondary education   
25     27         0     529.191635         28    bachelor's degree   
26     28         1     717.274324         26    bachelor's degree   
27     30         1  335581.668515         62  secondary education   
28     31         0    1682.083438         47    primary education   
29     32         0    4649.910832         34  secondary education   

    education_id      family_status  family_status_id gender income_type  \
0              0            married                 0      F    employee   
1              1            married                 0      F    employee   
2              1            married                 0      M    employee   
3              1            married                 0      M    employee   
4              1  civil partnership                 1      F     retiree   
5              0  civil partnership                 1      M    business   
6              0            married                 0      F    business   
7              1            married                 0      M    employee   
8              0  civil partnership                 1      F    employee   
9              1            married                 0      M    employee   
10             0            married                 0      M    business   
11             1            married                 0      F    employee   
12             2            married                 0      F    employee   
13             0  civil partnership                 1      F    business   
14             1            married                 0      F    employee   
15             1            married                 0      F    employee   
16             0  civil partnership                 1      M    employee   
17             1    widow / widower                 2      F     retiree   
18             1           divorced                 3      F    employee   
19             1            married                 0      M    employee   
20             1  civil partnership                 1      F    employee   
21             1  civil partnership                 1      F    employee   
22             0  civil partnership                 1      M    employee   
23       

In [41]:
# Look at the mean values for income based on your identified factors

mean_values_age = credit_scoring.groupby('age_categories')['total_income'].mean()
mean_values_age

age_categories
20's    25655.710253
30's    28395.528238
40's    28526.047215
50's    25725.431426
60's    23324.788869
70's    25945.907261
Teen    16993.942462
Name: total_income, dtype: float64

In [42]:
# Look at the median values for income based on your identified factors
median_values_age = credit_scoring.groupby('age_categories')['total_income'].median()
median_values_age

age_categories
20's    22903.6790
30's    24781.3960
40's    24701.9000
50's    22173.9030
60's    19764.1540
70's    22869.8235
Teen    14934.9010
Name: total_income, dtype: float64

In [43]:
# Mean based on education
mean_values_edu = credit_scoring.groupby('education')['total_income'].mean()
mean_values_edu

education
bachelor's degree      33142.802434
graduate degree        27960.024667
primary education      21144.882211
secondary education    24594.503037
some college           29040.132990
Name: total_income, dtype: float64

In [44]:
# Median based on education
median_values_edu = credit_scoring.groupby('education')['total_income'].median()
median_values_edu

education
bachelor's degree      28054.5310
graduate degree        25161.5835
primary education      18741.9760
secondary education    21836.5830
some college           25608.7945
Name: total_income, dtype: float64

In [45]:
# Mean based on income type
mean_values_income = credit_scoring.groupby('income_type')['total_income'].mean()
mean_values_income

income_type
business                       32386.741818
civil servant                  27343.729582
employee                       25820.841683
entrepreneur                   79866.103000
paternity / maternity leave     8612.661000
retiree                        21940.394503
student                        15712.260000
unemployed                     21014.360500
Name: total_income, dtype: float64

In [46]:
# Median based on income type
median_values_income = credit_scoring.groupby('income_type')['total_income'].median()
median_values_income

income_type
business                       27571.0825
civil servant                  24071.6695
employee                       22815.1035
entrepreneur                   79866.1030
paternity / maternity leave     8612.6610
retiree                        18962.3180
student                        15712.2600
unemployed                     21014.3605
Name: total_income, dtype: float64

In [47]:
# Mean based on number of children
mean_values_kids = credit_scoring.groupby('children')['total_income'].mean()
mean_values_kids

children
0    26421.916832
1    27368.627863
2    27478.854282
3    29322.623993
4    27289.829647
5    27268.847250
Name: total_income, dtype: float64

In [48]:
# Median based on number of children
median_values_kids = credit_scoring.groupby('children')['total_income'].median()
median_values_kids

children
0    23027.3350
1    23660.5630
2    23136.1155
3    25155.4480
4    24981.6340
5    29816.2255
Name: total_income, dtype: float64

Based on the data above, the education column seems to have the greatest effect on total income due to the range of the datapoints. Additionally, the median value seems to be most representative of the data considering there are many high datapoints in the 'total_income' category that are skewing the mean results.


In [49]:
credit_scoring['total_income'].describe()

count     19350.000000
mean      26787.266688
std       16475.822926
min        3306.762000
25%       16486.515250
50%       23201.873500
75%       32547.910750
max      362496.645000
Name: total_income, dtype: float64

In [50]:
#  Write a function that we will use for filling in missing values
sec_med_inc = credit_scoring[credit_scoring['education'] == 'secondary education']['total_income'].median()        
bac_med_inc = credit_scoring[credit_scoring['education'] == 'bachelor\'s degree']['total_income'].median()   
col_med_inc = credit_scoring[credit_scoring['education'] == 'some college']['total_income'].median()   
prim_med_inc = credit_scoring[credit_scoring['education'] == 'primary education']['total_income'].median()   
grad_med_inc = credit_scoring[credit_scoring['education'] == 'graduate degree']['total_income'].median()  

def median_income (row):
    if row['education'] == 'secondary education':
        return sec_med_inc
    if row['education'] == 'bachelor\'s degree':
        return bac_med_inc
    if row['education'] == 'some college':
        return col_med_inc
    if row['education'] == 'primary education':
        return prim_med_inc
    if row['education'] == 'graduate degree':
        return grad_med_inc
    return 'Other'

credit_scoring['median_income'] = credit_scoring.apply(lambda row: median_income(row), axis=1)

credit_scoring['total_income'].fillna(credit_scoring['median_income'], inplace = True)

In [51]:
# Check if it works
credit_scoring.head(15)

children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42    bachelor's degree             0   
1          1    4024.803754         36  secondary education             1   
2          0    5623.422610         33  secondary education             1   
3          3    4124.747207         32  secondary education             1   
4          0  340266.072047         53  secondary education             1   
5          0     926.185831         27    bachelor's degree             0   
6          0    2879.202052         43    bachelor's degree             0   
7          0     152.779569         50  secondary education             1   
8          2    6929.865299         35    bachelor's degree             0   
9          0    2188.756445         41  secondary education             1   
10         2    4171.483647         36    bachelor's degree             0   
11         0     792.701887         40  secondary education             1   
12         0            NaN         65  secondary education             1   
13         0    1846.641941         54         some college             2   
14         0    1844.956182         56    bachelor's degree             0   

        family_status  family_status_id gender income_type  debt  \
0             married                 0      F    employee     0   
1             married                 0      F    employee     0   
2             married                 0      M    employee     0   
3             married                 0      M    employee     0   
4   civil partnership                 1      F     retiree     0   
5   civil partnership                 1      M    business     0   
6             married                 0      F    business     0   
7             married                 0      M    employee     0   
8   civil partnership                 1      F    employee     0   
9             married                 0      M    employee     0   
10            married                 0      M    business     0   
11            married                 0      F    employee     0   
12  civil partnership                 1      M     retiree     0   
13            married                 0      F    employee     0   
14  civil partnership                 1      F    business     1   

    total_income                              purpose income_categories  \
0      40620.102                purchase of the house            40-45k   
1      17932.802                         car purchase            15-20k   
2      23341.752                purchase of the house            20-25k   
3      42820.568              supplementary education            40-45k   
4      25378.572                    to have a wedding            25-30k   
5      40922.170                purchase of the house            40-45k   
6      38484.156                 housing transactions            35-40k   
7      21731.829                            education            20-25k   
8      15337.093                     having a wedding            15-20k   
9      23108.150  purchase of the house for my family            20-25k   
10     18230.959                      buy real estate            15-20k   
11     12331.077           buy commercial real estate            10-15k   
12     21836.583                    to have a wedding          Over 60k   
13     20873.317                         car purchase            20-25k   
14     26420.466          buy residential real estate            25-30k   

   age_categories  median_income  
0            40's     28054.5310  
1            30's     21836.5830  
2            30's     21836.5830  
3            30's     21836.5830  
4            50's     21836.5830  
5            20's     28054.5310  
6            40's     28054.5310  
7            70's     21836.5830  
8            30's     28054.5310  
9            40's     21836.5830  
10           30's     28054.5310  
11           70's     21836.5830  
12           60's     2183

In [52]:
# Check if we got any errors
credit_scoring['total_income'].describe()

count     21453.000000
mean      26465.838089
std       15701.456362
min        3306.762000
25%       17219.352000
50%       22582.311000
75%       31327.922000
max      362496.645000
Name: total_income, dtype: float64

There are no errors. There are no missing values in the 'total_income' column now as the number of values matches the total number of rows in the dataset.

In [53]:
# Replacing missing values if there are any errors


In [54]:
# Checking the number of entries in the columns

credit_scoring.describe()

children  days_employed     dob_years  education_id  \
count  21453.000000   19350.000000  21453.000000  21453.000000   
mean       0.480585   66918.065141     43.474572      0.817042   
std        0.756079  139033.698578     12.213068      0.548628   
min        0.000000      24.141633     19.000000      0.000000   
25%        0.000000     926.990457     33.000000      1.000000   
50%        0.000000    2194.218768     43.000000      1.000000   
75%        1.000000    5538.423086     53.000000      1.000000   
max        5.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  median_income  
count      21453.000000  21453.000000   21453.000000    21453.00000  
mean           0.973896      0.081154   26465.838089    23449.14306  
std            1.421601      0.273078   15701.456362     2735.58389  
min            0.000000      0.000000    3306.762000    18741.97600  
25%            0.000000      0.000000   17219.352000    21836.58300  
50%            0.000000      0.000000   22582.311000    21836.58300  
75%            1.000000      0.000000   31327.922000    25608.79450  
max            4.000000      1.000000  362496.645000    28054.53100

###  Restoring values in `days_employed`

In [55]:
# Distribution of `days_employed` medians based on your identified parameters

median_days_inctype = credit_scoring.groupby('income_type')['days_employed'].median()
median_days_inctype


income_type
business                         1546.333214
civil servant                    2689.368353
employee                         1574.202821
entrepreneur                      520.848083
paternity / maternity leave      3296.759962
retiree                        365213.306266
student                           578.751554
unemployed                     366413.652744
Name: days_employed, dtype: float64

In [56]:
# median based on age

median_days_age = credit_scoring.groupby('age_categories')['days_employed'].median()
median_days_age

age_categories
20's      1013.438904
30's      1616.965312
40's      2153.248215
50's      5475.822930
60's    355956.740354
70's      2499.720678
Teen       724.492610
Name: days_employed, dtype: float64

In [57]:
# Distribution of `days_employed` means based on your identified parameters

mean_days_inctype = credit_scoring.groupby('income_type')['days_employed'].mean()
mean_days_inctype

income_type
business                         2111.470404
civil servant                    3399.896902
employee                         2326.499216
entrepreneur                      520.848083
paternity / maternity leave      3296.759962
retiree                        365003.491245
student                           578.751554
unemployed                     366413.652744
Name: days_employed, dtype: float64

In [58]:
# mean based on age

mean_days_age = credit_scoring.groupby('age_categories')['days_employed'].mean()
mean_days_age

age_categories
20's      2111.830222
30's      4411.185104
40's     14227.668212
50's    143178.856505
60's    287966.553964
70's     82809.857924
Teen       633.678086
Name: days_employed, dtype: float64

In this situation we will use age to fill in the missing values because it seems to have the most impact on days employed based on the range of values. We will use the mean values since there are no notable outliers and it seems to be the best representation of the data. 

In [59]:
# Let's write a function that calculates means or medians (depending on your decision) based on your identified parameter

teen_mean_days = credit_scoring[credit_scoring['age_categories'] == 'Teen']['days_employed'].mean()        
twenties_mean_days = credit_scoring[credit_scoring['age_categories'] == '20\'s']['days_employed'].mean()  
thirties_mean_days = credit_scoring[credit_scoring['age_categories'] == '30\'s']['days_employed'].mean()  
forties_mean_days = credit_scoring[credit_scoring['age_categories'] == '40\'s']['days_employed'].mean()  
fifties_mean_days = credit_scoring[credit_scoring['age_categories'] == '50\'s']['days_employed'].mean()  
sixties_mean_days = credit_scoring[credit_scoring['age_categories'] == '60\'s']['days_employed'].mean()  
seventies_mean_days = credit_scoring[credit_scoring['age_categories'] == '70\'s']['days_employed'].mean()  

def mean_days (row):
    if row['age_categories'] == 'Teen':
        return teen_mean_days
    if row['age_categories'] == '20\'s':
        return twenties_mean_days
    if row['age_categories'] == '30\'s':
        return thirties_mean_days
    if row['age_categories'] == '40\'s':
        return forties_mean_days
    if row['age_categories'] == '50\'s':
        return fifties_mean_days
    if row['age_categories'] == '60\'s':
        return sixties_mean_days
    return seventies_mean_days    

credit_scoring['mean_days'] = credit_scoring.apply(lambda row: mean_days(row), axis=1)

credit_scoring['days_employed'].fillna(credit_scoring['mean_days'], inplace = True)

In [60]:
# Check that the function works

credit_scoring.head(15)


children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42    bachelor's degree             0   
1          1    4024.803754         36  secondary education             1   
2          0    5623.422610         33  secondary education             1   
3          3    4124.747207         32  secondary education             1   
4          0  340266.072047         53  secondary education             1   
5          0     926.185831         27    bachelor's degree             0   
6          0    2879.202052         43    bachelor's degree             0   
7          0     152.779569         50  secondary education             1   
8          2    6929.865299         35    bachelor's degree             0   
9          0    2188.756445         41  secondary education             1   
10         2    4171.483647         36    bachelor's degree             0   
11         0     792.701887         40  secondary education             1   
12         0  287966.553964         65  secondary education             1   
13         0    1846.641941         54         some college             2   
14         0    1844.956182         56    bachelor's degree             0   

        family_status  family_status_id gender income_type  debt  \
0             married                 0      F    employee     0   
1             married                 0      F    employee     0   
2             married                 0      M    employee     0   
3             married                 0      M    employee     0   
4   civil partnership                 1      F     retiree     0   
5   civil partnership                 1      M    business     0   
6             married                 0      F    business     0   
7             married                 0      M    employee     0   
8   civil partnership                 1      F    employee     0   
9             married                 0      M    employee     0   
10            married                 0      M    business     0   
11            married                 0      F    employee     0   
12  civil partnership                 1      M     retiree     0   
13            married                 0      F    employee     0   
14  civil partnership                 1      F    business     1   

    total_income                              purpose income_categories  \
0      40620.102                purchase of the house            40-45k   
1      17932.802                         car purchase            15-20k   
2      23341.752                purchase of the house            20-25k   
3      42820.568              supplementary education            40-45k   
4      25378.572                    to have a wedding            25-30k   
5      40922.170                purchase of the house            40-45k   
6      38484.156                 housing transactions            35-40k   
7      21731.829                            education            20-25k   
8      15337.093                     having a wedding            15-20k   
9      23108.150  purchase of the house for my family            20-25k   
10     18230.959                      buy real estate            15-20k   
11     12331.077           buy commercial real estate            10-15k   
12     21836.583                    to have a wedding          Over 60k   
13     20873.317                         car purchase            20-25k   
14     26420.466          buy residential real estate            25-30k   

   age_categories  median_income      mean_days  
0            40's     28054.5310   14227.668212  
1            30's     21836.5830    4411.185104  
2            30's     21836.5830    4411.185104  
3            30's     21836.5830    4411.185104  
4            50's     21836.5830  143178.856505  
5            20's     28054.5310    2111.830222  
6            40's     28054.5310   14227.668212  
7            70's     21836.5830   82809.857924  
8            30's     28054.531

In [61]:
# Apply function to the income_type

credit_scoring['purpose'].value_counts()

wedding ceremony                            791
having a wedding                            768
to have a wedding                           765
real estate transactions                    675
buy commercial real estate                  661
housing transactions                        652
buying property for renting out             651
transactions with commercial real estate    650
purchase of the house                       646
housing                                     646
purchase of the house for my family         638
construction of own property                635
property                                    633
transactions with my real estate            627
building a real estate                      624
purchase of my own house                    620
buy real estate                             620
building a property                         619
housing renovation                          607
buy residential real estate                 606
buying my own car                       

In [62]:
credit_scoring = credit_scoring.replace(dict.fromkeys(['wedding ceremony','having a wedding','to have a wedding'], 'Wedding'))
credit_scoring = credit_scoring.replace(dict.fromkeys(['real estate transactions','buy commercial real estate','housing transactions','buying property for renting out','transactions with commercial real estate','housing','purchase of the house','purchase of the house for my family','construction of own property','property','transactions with my real estate','building a real estate','buy real estate','purchase of my own house','building a property','housing renovation','buy residential real estate'], 'Real Estate'))
credit_scoring = credit_scoring.replace(dict.fromkeys(['going to university','supplementary education','university education','education','to get a supplementary education','getting an education','profile education','getting higher education','to become educated'], 'Education'))
credit_scoring = credit_scoring.replace(dict.fromkeys(['buying my own car','car','second-hand car purchase','buying a second-hand car','to own a car','cars','to buy a car','car purchase','purchase of a car'], 'Car'))

In [63]:
# Check if function worked
credit_scoring['purpose'].value_counts()


Real Estate    10810
Car             4306
Education       4013
Wedding         2324
Name: purpose, dtype: int64

In [64]:
# Check the entries in all columns - make sure we fixed all missing values

credit_scoring.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 16 columns):
children             21453 non-null int64
days_employed        21453 non-null float64
dob_years            21453 non-null int64
education            21453 non-null object
education_id         21453 non-null int64
family_status        21453 non-null object
family_status_id     21453 non-null int64
gender               21453 non-null object
income_type          21453 non-null object
debt                 21453 non-null int64
total_income         21453 non-null float64
purpose              21453 non-null object
income_categories    21453 non-null object
age_categories       21453 non-null object
median_income        21453 non-null float64
mean_days            21453 non-null float64
dtypes: float64(4), int64(5), object(7)
memory usage: 2.6+ MB


All missing values have been addressed. Each column shows the same number of values (21453). 

## Checking the Hypotheses


**Is there a correlation between having children and paying back on time?**

In [81]:
# Check the children data and paying back on time

credit_scoring.groupby('children')['debt'].value_counts(normalize=True)*100

# Calculating default-rate based on the number of children

children  debt
0         0        92.455642
          1         7.544358
1         0        90.834192
          1         9.165808
2         0        90.507519
          1         9.492481
3         0        91.818182
          1         8.181818
4         0        90.243902
          1         9.756098
5         0       100.000000
Name: debt, dtype: float64

**Conclusion**

Those without children are the least likely to default on a loan, with a 7.5% default rate. Although the data indicates that 100% of those with 5 children paid their loan on time, this is not necessarily a valid conclusion due to the small sample size of those with 5 children (9). Likelihood of defaulting on a loan typically increases as number of children increase, although the likelihood decreases slightly for those with 3 children.

**Is there a correlation between family status and paying back on time?**

In [80]:
# Check the family status data and paying back on time

credit_scoring.groupby('family_status')['debt'].value_counts(normalize=True)*100

# Calculating default-rate based on family status



family_status      debt
civil partnership  0       90.650602
                   1        9.349398
divorced           0       92.887029
                   1        7.112971
married            0       92.454818
                   1        7.545182
unmarried          0       90.249110
                   1        9.750890
widow / widower    0       93.430657
                   1        6.569343
Name: debt, dtype: float64

**Conclusion**

Those that are unmarried are most likely to default on a loan (9.75%) while those that are widowers are least likely to default (6.5%).

**Is there a correlation between income level and paying back on time?**

In [79]:
# Check the income level data and paying back on time

credit_scoring.groupby('income_categories')['debt'].value_counts(normalize=True)*100

# Calculating default-rate based on income level



income_categories  debt
0-5k               0       92.307692
                   1        7.692308
10-15k             0       91.480298
                   1        8.519702
15-20k             0       91.450634
                   1        8.549366
20-25k             0       91.476768
                   1        8.523232
25-30k             0       91.095380
                   1        8.904620
30-35k             0       92.108108
                   1        7.891892
35-40k             0       92.356688
                   1        7.643312
40-45k             0       93.494229
                   1        6.505771
45-50k             0       92.578850
                   1        7.421150
5-10k              0       93.777778
                   1        6.222222
50-55k             0       92.207792
                   1        7.792208
55-60k             0       90.874525
                   1        9.125475
Over 60k           0       92.504505
                   1        7.495495
Name: debt, dt

**Conclusion**

Very low-income borrowers (0-10k) seem to have low rates of defaulting on loans. The likelihood of defaulting increases for those between 10-30k. Those making 30-55k and over 60k have similar likelihoods of defaulting. The highest chance of defaulting is for incomes between 55-60k (9.1%).

**How does credit purpose affect the default rate?**

In [76]:
# Check the percentages for default rate for each credit purpose and analyze them
credit_scoring.groupby('purpose')['debt'].value_counts(normalize=True)*100


purpose      debt
Car          0       90.640966
             1        9.359034
Education    0       90.779965
             1        9.220035
Real Estate  0       92.765957
             1        7.234043
Wedding      0       91.996558
             1        8.003442
Name: debt, dtype: float64

**Conclusion**

According to this dataset, the two categories of purpose that are most likely to default on a loan are for car purchases (9.4%) and education (9.2%). Weddings and real estate purchases follow closely behind with default rates of 8.0% and 7.2% respectively. 

# General Conclusion 

To answer the primary concerns of this report, the marital status and number of children does have an impact on the likelihood of defaulting on a loan. Generally speaking, the likelihood of defaulting on a loan tends to increase with the number of children the borrower has. Additionally, unmarried borrowers are the most likely to default on a loan while widow/widowers are the least likely. Other factors that can affect the likelihood of defaulting a loan are the purpose of the loan and the income level of the borrower. Loans being used for car purchases have the highest chance of defaulting while those that are used for real estate purchases have the lowest chance. Furthermore, lowest income borrowers actually have the lowest likelihood of defaulting on a loan while those in the $55-60k category has the highest chance. 

There were many missing data points in the 'days_employed' and 'total_income' categories. The 'days_employed' values were replaced using the mean of all the values in a particular age range. Age was the chosen category in this situation due to the varying range of value when grouped by age. Mean value was used because there were no significant outliers. For total_income missing values were replaced based on the median values when grouped by education level. This category was also selected because of the high range of values when grouped by education. In this case, the median values were used as there were several significant outliers. 

There were also a sizeably portion of the data that were duplicates. These duplicates made up a relatively low percentage of the total data and they were dropped.